### ISSUE TITLE
INTL_IDNA_VARIANT_UTS46 not defined having ICU < 4.6

### ISSUE THREAD

**Issue by ohader** [guzzle/guzzle#2448]
**Guzzle version(s) affected**: 6.5.0

**Description**  
`IDN conversion failed` exception being thrown when dealing with domain name having regular ASCII characters (no IDN conversion required at all). Exception is thrown at https://github.com/guzzle/guzzle/blob/6.5.0/src/Client.php#L241

**How to reproduce**  
Requires old version of ICU on e.g. PHP 7.2 (these two things are independent - unfortunately).

**Possible Solution** 
Omit `INTL_IDNA_VARIANT_UTS46` in case it's not available - at https://github.com/guzzle/guzzle/blob/6.5.0/src/Client.php#L221

**Additional References**

See the following references and how it was handled in the TYPO3 CMS project
* https://github.com/TYPO3/TYPO3.CMS/commit/3a2b54a1b36b99b525cc04eccf20fd018a24e4a6#diff-cb4f296a39c2f11c2784e0e940745b82R193-R199
* https://forge.typo3.org/issues/87953 (additional description & system details)

**Comment by exotec** [comment]
I also had problems with it this morning. Downgrade to 6.4.1 works

**Comment by ohader** [comment]
> ICU version => 51.1

I'm not entirely sure how ICU's versioning is supposed to work - referring to https://github.com/php/php-src/blob/PHP-7.2/ext/intl/idn/idn.c#L37 (same in current PHP 7.3 branch) they related to `ICU 4.6`

@exotec What was the exact PHP version (and vendor OS) you're using?

**Comment by exotec** [comment]
PHP 7.2.10 
Linux s1036 4.19.44-gentoo-mw1

**Comment by bramus** [comment]
Problem was introduced with commit https://github.com/guzzle/guzzle/commit/9d4716949a431e55752c7aa37d29566affcd4037#diff-ff73e042e738204c6da009e2ed19f783 (and shipped with 6.5.0)

Other libraries that use it include checks for its existence, falling back to `INTL_IDNA_VARIANT_2003`. See https://github.com/bcit-ci/CodeIgniter/commit/4541bf930e715917622dc2d225bf5517f99db35a for example.

Please release a hotfix. All CentOS 6 servers (on which many cPanel servers run) for example are stuck with ICU 4.2.1 and thus are affected by this.

**Comment by ohader** [comment]
> Other libraries that use it include checks for its existence, falling back to `INTL_IDNA_VARIANT_2003`. See [bcit-ci/CodeIgniter@4541bf9](https://github.com/bcit-ci/CodeIgniter/commit/4541bf930e715917622dc2d225bf5517f99db35a) for example.

Back then we had a similar idea, however on PHP 7.2 this leads to a corresponding deprecation message. That's why we kept the default option if `UTS46` is not available - see
* https://forge.typo3.org/issues/87953#note-14
* https://www.php.net/manual/en/function.idn-to-ascii.php#refsect1-function.idn-to-ascii-changelog

**Comment by alexeyshockov** [comment]
@ohader, @bramus, @exotec, please check #2453. Hope I didn't forget something.

**Comment by jonnott** [comment]
+1 for a fix .. I have CentOS 6 with ICU 4.2.1

**Comment by imajinyun** [comment]
Skip version 6.5.0 and you're ready to go

```bash
composer require guzzlehttp/guzzle:'>=6.3.0 <6.5.0 || ^6.5.1'
```

**Comment by Nyholm** [comment]
This is fixed by https://github.com/guzzle/guzzle/pull/2454

Will be released in 6.5.1. (hopefully today)

**Comment by jdmendez** [comment]
I might not understand the problem with this issue.  I used the composer solution and it added a bunch of weird new scripts that I way too unnecessary.  I solve the issue by commenting out the line 1153 where the constant is used, and defined the variant as null.  That solved my problem without increasing risk with new scripts.
